### PR TITLE
Add capability to compute average of a variable over the exposed portion of a boundary

### LINKED ISSUE #23949: Add ability to compute average of variable over exposed boundary
## Reason
Issue #23097 added a capability to the HeatConduction module to determine which portions of a surface are shadowed by portions of the physical model. To supplement that, we need the ability to compute the average of a variable (such as temperature) over just the portion of the boundary that is exposed.

## Design
Develop a PostProcessor that derives from SideAverageValue, but which requires that the user supply the name of a SelfShadowSideUserObject, and use that to determine whether a QP is shadowed or exposed. If it is exposed, include its contribution to both the area integral and the volume used by SideAverageValue.

## Impact
This will only add new capability.

### PR DESCRIPTION
**Pull request by bwspenc**
closes #23949

### CODE REVIEW

File: modules/heat_conduction/include/postprocessors/ExposedSideAverageValue.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  Real computeQpIntegral() override;
+
+protected:
+  /**
+   * Instead of computing the total face volume (as the base class does), compute
+   * the volume of the exposed QPs on that face.
+   * @return exposed face volume
+   */
+  virtual Real volume() override;
+  /// Reference to SelfShadowSideUserObject, which does the determination of which
```
Comment (lindsayad) [inline modules/heat_conduction/include/postprocessors/ExposedSideAverageValue.h:35]:
```suggestion

  /// Reference to SelfShadowSideUserObject, which does the determination of which
```

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      "SelfShadowSideUserObject that calculates the illumination state of a side set");
+  return params;
+}
+
+ExposedSideAverageValue::ExposedSideAverageValue(const InputParameters & parameters)
+  : SideAverageValue(parameters),
+    _self_shadow(getUserObject<SelfShadowSideUserObject>("self_shadow_uo"))
+{
+  if (_self_shadow.useDisplacedMesh() != getParam<bool>("use_displaced_mesh"))
+    paramWarning("use_displaced_mesh",
```
Comment (lindsayad) [inline modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:33]: Should this be an error?

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    return 0.0;
+}
+
+Real
+ExposedSideAverageValue::volume()
+{
+  Real curr_exposed_side_volume = 0.0;
+  const SelfShadowSideUserObject::SideIDType id(_current_elem->id(), _current_side);
+  const int illumination = _self_shadow.illumination(id);
+  for (unsigned int qp = 0; qp < _qrule->n_points(); qp++)
```
Comment (lindsayad) [inline modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:56]:
```suggestion
  for (const unsigned int qp : make_range(_qrule->n_points()))
```

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  {
+    // tests if the bit at position _qp is set
+    if (illumination & (1 << qp))
```
Comment (lindsayad) [inline modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:59]: https://wiki.sei.cmu.edu/confluence/display/c/INT13-C.+Use+bitwise+operators+only+on+unsigned+operands suggests bitwise operations should only ever be performed on unsigned integers

**Review by lindsayad — CHANGES_REQUESTED** [review]
Seems like we should be updating some requirements for the tests you're modifying?

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C (hunk 2 above)
Comment (bwspenc) [reply modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C]: Yes -- changed it to an error.

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C (hunk 4 above)
Comment (bwspenc) [reply modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:59]: Agreed. I changed `SelfShadowSideUserObject::illumination()` to return an unsigned int, and changed the variable types accordingly.

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+registerMooseObject("MooseApp", ExposedSideAverageValue);
+
+InputParameters
+ExposedSideAverageValue::validParams()
+{
+  InputParameters params = SideAverageValue::validParams();
+  params.addClassDescription("Computes the average value of a variable on the "
+                             "exposed portion of a sideset. Note that this cannot be used on the "
+                             "centerline of an axisymmetric model.");
```
[the remark this answers is not in the document]
Comment (bwspenc) [reply modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:21]: This class derives from `SideAverageValue` (in `framework/src/postprocessors`), from which I carried over the same note in its class description. The actual integration over the sides is done in `SideIntegralPostprocessor::computeIngegral()`. I believe that if a qp is on the centerline, its value of `_coord` will be zero. If this happens, we'll contribute zero to the integral (which is the correct thing to do), and `SideAverageValue::volume()` will return zero. If we only have faces on the centerline, we would get a divide by zero in `SideAverageValue::getValue(), which I guess we should check against. Maybe we should talk for a few minutes about how to handle this.

[the remark this answers is not in the document]
Comment (bwspenc) [reply modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:21]: I added a check for the divide by zero mentioned above.

**Review by lindsayad — DISMISSED** [review]
Looks good @bwspenc ! Merge when you're ready (if tests pass 😄 )

File: modules/heat_conduction/doc/content/source/postprocessors/ExposedSideAverageValue.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# ExposedSideAverageValue
+
+!syntax description /Postprocessors/ExposedSideAverageValue
+
+This is used to compute the average of a variable over only the portion of a specified surface that is exposed, using the [SelfShadowSideUserObject](SelfShadowSideUserObject.md) to determine which portions of the surface are exposed and which are shadowed. This computation is identical to that performed in [SideAverageUserObject](SideAverageUserObject.md), except that only portions of the surface that are exposed (i.e., not shadowed) are included in the calculation.
+
+!listing directional_flux_bc/2d.i block=Postprocessors/ave_v_exposed
```
Comment (GiudGiud) [inline modules/heat_conduction/doc/content/source/postprocessors/ExposedSideAverageValue.md:7]: This is not useful unless you add text to say what s being done.
The !syntax inputs /Postprocessors/ExposedSideAverageValue lists this input already

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ExposedSideAverageValue.h"
+#include "SelfShadowSideUserObject.h"
+
+registerMooseObject("MooseApp", ExposedSideAverageValue);
```
Comment (GiudGiud) [inline modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:13]: Not the right app

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+               "undisplaced) as this PostProcessor.");
+}
+
+Real
+ExposedSideAverageValue::computeQpIntegral()
+{
+  const SelfShadowSideUserObject::SideIDType id(_current_elem->id(), _current_side);
+  const unsigned int illumination = _self_shadow.illumination(id);
+  // tests if the bit at position _qp is set
+  if (illumination & (1 << _qp))
```
Comment (GiudGiud) [inline modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:44]: this `1 << _qp` thing does not seem right. where did it come from? would you have a minute to explain it to me?

File: test/tests/userobjects/side_user_object_no_boundary_error/tests [hunk 9]
```diff
@@ -26,4 +26,13 @@
...
     requirement = "The system shall error if the lower-d variable is not properly defined on a boundary."
   []
+
+  [rz_centerline_side]
+    type = RunException
+    input = side_no_boundary.i
+    cli_args = "Postprocessors/avg/boundary='left' Problem/coord_type=RZ"
+    mesh_mode = replicated
+    expect_err = "The total area of the boundary is zero. This could be due to using a boundary on the centerline of an axisymmetric model."
+    requirement = "The system shall error if SideAverageValue is applied to a boundary on the centerline of an axisymmetric model."
```
Comment (GiudGiud) [inline test/tests/userobjects/side_user_object_no_boundary_error/tests:36]:
```suggestion
    requirement = "The system shall error if the side average value of a variable is requested on a boundary that is the centerline of an axisymmetric model."
```

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+InputParameters
+ExposedSideAverageValue::validParams()
+{
+  InputParameters params = SideAverageValue::validParams();
+  params.addClassDescription("Computes the average value of a variable on the "
+                             "exposed portion of a sideset. Note that this cannot be used on the "
+                             "centerline of an axisymmetric model.");
+  params.addRequiredParam<UserObjectName>(
+      "self_shadow_uo",
+      "SelfShadowSideUserObject that calculates the illumination state of a side set");
```
Comment (GiudGiud) [inline modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:24]:
```suggestion
      "SelfShadowSideUserObject that calculates the illumination state of element sides in a side set");
```

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C (hunk 8 above)
Comment (dschwen) [reply modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:44]: It is right. `<<` is the left shift operator. It generates the values 1 2 4 8 for _qp 0 1 2 3 (which are bit masks to test bit 0 1 2 3). The illumination state of all QPs of a side element is stored as a bitmask.

File: modules/heat_conduction/doc/content/source/postprocessors/ExposedSideAverageValue.md (hunk 6 above)
Comment (bwspenc) [reply modules/heat_conduction/doc/content/source/postprocessors/ExposedSideAverageValue.md]: Removed

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C (hunk 7 above)
Comment (bwspenc) [reply modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C]: Good catch! I fixed that.

File: test/tests/userobjects/side_user_object_no_boundary_error/tests (hunk 9 above)
Comment (bwspenc) [reply test/tests/userobjects/side_user_object_no_boundary_error/tests]: Applied changes

File: modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C (hunk 8 above)
Comment (bwspenc) [reply modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:44]: I believe this is implemented correctly like @dschwen commented. However, I also think this is a confusing way to get that information from `SelfShadowSideUserObject`. I didn't really want to touch the implementation of `SelfShadowSideUserObject` in this PR. Maybe that could be improved in a follow-on PR?

**Review by GiudGiud — DISMISSED** [review]
bitwise approval

File: modules/heat_conduction/doc/content/source/postprocessors/ExposedSideAverageValue.md [hunk 11]
```diff
@@ -0,0 +1,11 @@
+# ExposedSideAverageValue
+
+!syntax description /Postprocessors/ExposedSideAverageValue
+
+This is used to compute the average of a variable over only the portion of a specified surface that is exposed, using the [SelfShadowSideUserObject](SelfShadowSideUserObject.md) to determine which portions of the surface are exposed and which are shadowed. This computation is identical to that performed in [SideAverageUserObject](SideAverageUserObject.md), except that only portions of the surface that are exposed (i.e., not shadowed) are included in the calculation.
```
Comment (lindsayad) [inline modules/heat_conduction/doc/content/source/postprocessors/ExposedSideAverageValue.md:5]:
```suggestion
This is used to compute the average of a variable over only the portion of a specified surface that is exposed, using the [SelfShadowSideUserObject](SelfShadowSideUserObject.md) to determine which portions of the surface are exposed and which are shadowed. This computation is identical to that performed in [SideAverageValue](SideAverageValue.md), except that only portions of the surface that are exposed (i.e., not shadowed) are included in the calculation.
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: modules/heat_conduction/doc/content/source/postprocessors/ExposedSideAverageValue.md (hunk 11 above)
Comment (bwspenc) [reply modules/heat_conduction/doc/content/source/postprocessors/ExposedSideAverageValue.md]: Fixed. Thanks for finding this issue!

**Review by lindsayad — APPROVED** [review]

**Pull request merged**